service/storage_service.cc

```diff
@@ -2338,15 +2338,16 @@ void storage_service::excise(std::unordered_set<token> tokens, inet_address endp
     slogger.info("Removing tokens {} for {}", tokens, endpoint);
     // FIXME: HintedHandOffManager.instance.deleteHintsForEndpoint(endpoint);
     remove_endpoint(endpoint);
-    auto tmlock = get_token_metadata_lock().get0();
+    auto tmlock = std::make_optional(get_token_metadata_lock().get0());
```

Comment:
nitpick: instead of creating an optional, it would be more readable for me personally if you put this block of code inside a separate scope (`{ }`). As a bonus, you'd also very slightly reduce the scope of `tmptr`.

Reply:
When I was reviewing this I had a similar thought but then I came to the conclusion that it makes absolutely no difference whether it's done this way or the other. I personally find additional nesting level less readable ;).

Reply:
I considered doing this, but:

1. using an optional results in a smaller patch - I'd have to add a `{ }` scope in one commit and then indent in the other,
2. a similar pattern is already used in `storage_service.cc` - look at the `tmlock` variable in `handle_state_normal` and `prepare_to_join` functions.

Reply:
It's alright, I added a `nitpick` prefix and `for me personally` postfix to indicate that I don't deeply care one way or the other :) Let's wait for a reasonable hour in Vlad's time zone to give him a chance to review this code and I'm fine with merging afterwards.

Reply:
(also, since this topic came up: it's absolutely fine to not split such small changes into code + indentation, especially if you're posting a github pull request instead of a mailing list patch)

Reply:
Could you, please, explain to me why this change needed at all?

AFAIKT having a single liner
```
+       tmlock.reset();
```

as below should have been sufficient.
What do I miss here?

Reply:
Are you looking at existing code like:

 348     // Replicate the tokens early because once gossip runs other nodes
 349     // might send reads/writes to this node. Replicate it early to make
 350     // sure the tokens are valid on all the shards.
 351     replicate_to_all_cores(std::move(tmptr)).get();
 352     tmlock.reset();

Here tmlock is defined as:

auto tmlock = std::make_unique<token_metadata_lock>(get_token_metadata_lock().get0());

There is no token_metadata_lock::reset()

Reply:
I wanted to stick to a convention I observed in `storage_service` and rely on `token_metadata_lock`'s destructor to release it. Having an optional and resetting it at some point lets me achieve that.

There are other places in `storage_service` which use a similar pattern - look at the `tmlock` variable in `handle_state_normal` and `prepare_to_join` methods. Although they use `unique_ptr` instead of `optional`, the idea is the same.

To be fair, `token_metadata_lock` is a typedef for `semaphore_units`, so I could have called `return_all` and achieved the same effect without introducing an optional. However, I think relying on the destructor is better, as we can treat the lock as an opaque object. Even if we change the underlying type of `token_metadata_lock`, as long as it releases the lock on destruction, the code will be future-proof. On the other hand, the new underlying type might not have a `return_all` method at all.

Reply:
If this lock is the only concern then I'll enqueue this series (even though I still like scopes more), especially that compiler seem to be perfectly capable to optimize out such use of optionals: https://godbolt.org/z/q1n9fWPWa

Reply:
> I wanted to stick to a convention I observed in storage_service and rely on token_metadata_lock's destructor to release it. Having an optional and resetting it at some point lets me achieve that.

@piodul I don't understand why do you need to wrap it into an `optional`? 
Your answer implies that 
```
auto tmlock = get_token_metadata_lock().get0();
```
would not lead to call of a destructor in question when you are going to leave the scope and I can't get why. I sense I'm missing something here.

AFAIU all this wrapping is not supposed to be needed. The same about calling to `semaphore_units::reset()` - should not be needed.

Could you, please, clarify?

Reply:
@psarna @avikivity @slivne It's funny how patches are getting merged even thought there are non-cosmetics comments and requests for a change. This makes the review procedure pointless.

If this becomes a common practice I suggest to limit the amount of people tagged on particular PRs only to those who's comments ARE going to be accounted because reviewing consumes a lot of time and effort.

Reply:
> > I wanted to stick to a convention I observed in storage_service and rely on token_metadata_lock's destructor to release it. Having an optional and resetting it at some point lets me achieve that.
> 
> @piodul I don't understand why do you need to wrap it into an `optional`?
> Your answer implies that
> 
> ```
> auto tmlock = get_token_metadata_lock().get0();
> ```
> 
> would not lead to call of a destructor in question when you are going to leave the scope and I can't get why. I sense I'm missing something here.

If I didn't use an optional or didn't do `.reset()` then yes, the lock would be freed in the destructor at the end of the scope. However, I wanted it to be destroyed earlier. After the line:

```c++
replicate_to_all_cores(std::move(tmptr)).get();
```

the token metadata is fully updated, so there is no point in holding the lock further. There is no need to hold it while `notify_left` is called as it invokes callbacks which could potentially be long-running.

Reply:
@vladzcloudius the review process isn't pointless and each comment is valuable - if something's missing, it can always be provided as a follow-up. But in this case the request *was* cosmetic, if I understand the discussion above correctly, so it's good that Tomek enqueued the series as it reached master faster - in the end we just want the lock to be released at specific time, and the exact technique to achieve that is not critical for this series to be merged. As I said, I'd personally prefer a new scope rather than an optional, but it's not a blocker.

Reply:
I wasn't asking about "how", I was asking about "why" - this is not cosmetics at all.
The outcome of this discussion is irrelevant because next time I could have been right - that's the whole point of a review.

If you think that you don't need my opinion on this series and "know better" which seems to be the case - then, again, no need to ask for my review.

Reply:
> If I didn't use an optional or didn't do `.reset()` then yes, the lock would be freed in the destructor at the end of the scope. However, I wanted it to be destroyed earlier. After the line:
> 
> ```c++
> replicate_to_all_cores(std::move(tmptr)).get();
> ```
> 
> the token metadata is fully updated, so there is no point in holding the lock further. There is no need to hold it while `notify_left` is called as it invokes callbacks which could potentially be long-running.

Oh, I finally got it. 
Thanks for a clarification, @piodul.